### ISSUE TITLE
Dataset object is operatable even after closing the checkout

### ISSUE THREAD

**Issue by hhsecond** [tensorwerk/hangar-py#36]
## Expected Behavior
After closing the checkout, the dataset object should not be operatable. We need to decide what all are possible on the dataset object though. So far, we know
1. User should not be able to add data.
2. User should not be able to fetch data

## Current Behavior
1.
```python
co = repo.checkout(write=True)
dset = co.datasets.init_dataset('dset', prototype=array5by7)
co.commit()
co.close()
co = repo.checkout(write=True)  # without opening the checkout again, user won't be able to do the addition even now
dset.add(array5by7, '1')  # user can add data to dset object
```
2.
```python
co = repo.checkout(write=True)
dset = co.datasets.init_dataset('dset', prototype=array5by7)
dset.add(array5by7, '1')
co.commit()
co.close()
print(dset['1'])  # user can fetch data from dset object
```

These issues are part of #25 
- https://github.com/tensorwerk/hangar-py/pull/25#discussion_r281058557
- https://github.com/tensorwerk/hangar-py/pull/25#discussion_r281058815

**Comment by rlizzo** [comment]
So the problem with this is that it's stupidly simple to fix, but adds more the the increasing list of verifications which need to occur in order to get or add data to a dataset/metadata object... we need some benchmarks soon before we "verify" ourselves into a tight spot here. 

Note: Most of the explanation below is for `write-enabled` checkouts, since that is the really dangerous case where this could go wrong, a `read only` checkout could only ever just act odd, not actually cause harm to the underlying records. 

Since a user is provided direct access to the dataset objects, and are allowed to keep references to them outside the hangar namespace (and checkout object from which they are constructed), closing a checkout currently only `"self destructs"` the ability for new `dataset`/`metadata` object access. See:

https://github.com/tensorwerk/hangar-py/blob/065ae0735c630f627fb5f5d9cd634739408a5733/src/hangar/checkout.py#L158-L177

Since the repo object has no way to communicate with reviously created `dataset` or `metadata` object to tell them they need to stop working, the fix would be to implement checks in the object themselves. By passing the `__writer_lock` key and `_branchenv` environment down the stack, the objects could in theory ask for permission to operate every time they want to do something (like the checkout object does). 

It's simple but will cost time and may introduce more overhead then we would like. I don't think this cost is so trivial because I think (at a glance) that we will need to open and close the `_branchenv` `transaction handle` on every call so that updates are immediatly seen. This absolutely murders performance (literally dropping in half or more for interacting with smaller sample sizes). 

However, it does need to be fixed. 

Any suggestions @hhsecond ?

**Comment by rlizzo** [comment]
Also pinging @lantiga since this is fairly critical

**Comment by rlizzo** [comment]
Ok. So I'm wondering if there isn't a way to invalidate any calls to the underlying `lmdb transactions` when a checkout is closed instead of explicitly doing a verification call each time. There may be hope here

**Comment by rlizzo** [comment]
> Ok. So I'm wondering if there isn't a way to invalidate any calls to the underlying `lmdb transactions` when a checkout is closed instead of explicitly doing a verification call each time. Just catching an exception and using that as our signal to destruct the object would be much more performant
> 
> There may be hope yet...

This is actually much more convoluted then it needs to be, `weakref proxies` solve this problem really elegantly. 

https://docs.python.org/3/library/weakref.html

I have a prototype now, should have the fix pushed in the next few hours

**Comment by rlizzo** [comment]
closed in #41
